### ISSUE TITLE
Add action name detection for DelayedJob::Structs

### ISSUE THREAD

**Issue by tombruijn** [appsignal/appsignal-ruby#637]
Currently apps need to define the `appsignal_name` method to get an action name to AppSignal for a job that's built this way.

But I can build in the plugin a valid action name based on the payload:

```
> "#{payload.object.class.name}##{payload.method_name}"
=> "WelcomeMailStructJob#deliver_without_delay"
```

Based on this example app: https://github.com/appsignal/appsignal-examples/tree/rails-5+delayed_job

Let's add support for this method to the Delayed::Job plugin.

---

- https://app.intercom.com/a/apps/yzor8gyw/inbox/inbox/conversation/16410700001487

**Comment by tombruijn** [comment]
My testing shows that it already picks it up on a struct job on the latest Ruby gem version.

**Comment by wok** [comment]
Tested with `appsignal 2.10.10` and `delayed_job 4.1.8`, still getting **unknown** reported to AppSignal UI.

When I query the job name from Rails console I get the following:
```
2.6.3 :002 > Delayed::Job.last.name
  Delayed::Backend::ActiveRecord::Job Load (3.4ms)  SELECT  `delayed_jobs`.* FROM `delayed_jobs` ORDER BY `delayed_jobs`.`id` DESC LIMIT 1
 => "Invoice::DeliveryJob"
```
And then [AppSignal::Hooks::DelayedJobPlugin#class_and_method_name_from_object_or_hash](https://github.com/appsignal/appsignal-ruby/blob/main/lib/appsignal/integrations/delayed_job_plugin.rb#L62) will ignore the job name since it cannot be split into two parts.

`Invoice::DeliveryJob` looks like this:
```
class DeliveryJob < Struct.new(:invoice_id)
  def perform
    invoice = Invoice.find(invoice_id)
    DefaultMailer.invoice(invoice).deliver_now
  end
end
```

**Comment by tombruijn** [comment]
Ah thanks! I think the key here is how it's enqueued, with "Delayed::Job.enqueue".

Time to really rewrite the DJ integration to not use monitor_transaction.

**Comment by tombruijn** [comment]
Hi @wok, I've released a fix in `2.11.0.beta.4` of the gem. Can you test it in a staging/acceptance environment? And let me know if it works or not.

Closing for now, but feel free to reply here and mention me `@tombruijn`

**Comment by wok** [comment]
Hi @tombruijn,

just tested in development environment, and I can confirm that the action name is now reported correctly to AppSignal.

Maybe the final return could be changed to return an array, e.g.

```
[default_name || 'unknown', 'perform']
```

The [caller](https://github.com/appsignal/appsignal-ruby/blob/30f750c80983b11e48b807fa4645889cc0866fae/lib/appsignal/integrations/delayed_job_plugin.rb#L31) expects an array where the first element is ``class_name`` and the second ``method_name``. Now ``#{default_name}#perform`` is reported as `class_name`.

Nothing major though.

**Comment by tombruijn** [comment]
Hi @wok, It shouldn't be necessary here. I can't find a scenario that returns `""` for a `job.name`.
Other cases in that method already return single values for the class method action name. So no need to return an array at all times. The `method_name` value will be empty and the action name builder in the transaction can handle that scenario.

I had a commit ready that rewrote the entire thing to not rely on `monitor_transaction` and allow us to set the action name in one part instead of having to split it up. But that's too big to apply at once in a PR, so if I'm going to apply that I'll merge it smaller PRs. Instead I took out only the part that was necessary for the fix in #642
